File: sdks/javascript/src/resources/campaigns.ts
/**
 * Campaigns resource for managing marketing campaigns
 */

import type { Mailat } from '../client'
import type {
  Campaign,
  CreateCampaignOptions,
  UpdateCampaignOptions,
  CampaignStats,
  PaginatedResponse
} from '../types'

export class CampaignsResource {
  constructor(private client: Mailat) {}

  /**
   * Create a new campaign
   *
   * @example
   * ```typescript
   * const campaign = await client.campaigns.create({
   *   name: 'Summer Newsletter',
   *   subject: 'Summer is here! Check out our deals',
   *   listIds: ['newsletter-subscribers'],
   *   fromName: 'Acme Inc',
   *   fromEmail: 'marketing@acme.com',
   *   htmlContent: '<h1>Summer Deals</h1>...'
   * })
   * ```
   */
  async create(options: CreateCampaignOptions): Promise<Campaign> {
    return this.client.post<Campaign>('/campaigns', options)
  }

  /**
   * Get campaign by ID
   */
  async get(id: string): Promise<Campaign> {
    return this.client.get<Campaign>(`/campaigns/${id}`)
  }

  /**
   * Update a campaign (only draft campaigns can be updated)
   */
  async update(id: string, options: UpdateCampaignOptions): Promise<Campaign> {
    return this.client.put<Campaign>(`/campaigns/${id}`, options)
  }

  /**
   * Delete a campaign
   */
  async delete(id: string): Promise<void> {
    await this.client.delete(`/campaigns/${id}`)
  }

  /**
   * List campaigns
   */
  async list(options: {
    page?: number
    limit?: number
    status?: string
  } = {}): Promise<PaginatedResponse<Campaign>> {
    return this.client.get<PaginatedResponse<Campaign>>('/campaigns', {
      page: options.page,
      limit: options.limit,
      status: options.status
    })
  }

  /**
   * Send campaign immediately
   */
  async send(id: string): Promise<Campaign> {
    return this.client.post<Campaign>(`/campaigns/${id}/send`)
  }

  /**
   * Schedule campaign for future sending
   *
   * @example
   * ```typescript
   * const campaign = await client.campaigns.schedule('campaign-id', {
   *   scheduledAt: new Date('2024-12-25T09:00:00Z')
   * })
   * ```
   */
  async schedule(id: string, options: { scheduledAt: Date | string }): Promise<Campaign> {
    return this.client.post<Campaign>(`/campaigns/${id}/schedule`, {
      scheduledAt: new Date(options.scheduledAt).toISOString()
    })
  }

  /**
   * Pause a sending campaign
   */
  async pause(id: string): Promise<Campaign> {
    return this.client.post<Campaign>(`/campaigns/${id}/pause`)
  }

  /**
   * Resume a paused campaign
   */
  async resume(id: string): Promise<Campaign> {
    return this.client.post<Campaign>(`/campaigns/${id}/resume`)
  }

  /**
   * Cancel a scheduled or sending campaign
   */
  async cancel(id: string): Promise<Campaign> {
    return this.client.post<Campaign>(`/campaigns/${id}/cancel`)
  }

  /**
   * Get campaign statistics
   */
  async getStats(id: string): Promise<CampaignStats> {
    return this.client.get<CampaignStats>(`/campaigns/${id}/stats`)
  }

  /**
   * Preview campaign HTML
   */
  async preview(id: string, contactId?: string): Promise<{ html: string; subject: string }> {
    return this.client.post(`/campaigns/${id}/preview`, { contactId })
  }

  /**
   * Send test email for campaign
   *
   * @example
   * ```typescript
   * await client.campaigns.sendTest('campaign-id', ['test@example.com'])
   * ```
   */
  async sendTest(id: string, emails: string[]): Promise<{ sent: number }> {
    return this.client.post(`/campaigns/${id}/test`, { emails })
  }

  /**
   * Duplicate a campaign
   */
  async duplicate(id: string, name?: string): Promise<Campaign> {
    return this.client.post<Campaign>(`/campaigns/${id}/duplicate`, { name })
  }

  /**
   * Get campaign activity feed
   */
  async getActivity(id: string, options: {
    page?: number
    limit?: number
    eventType?: string
  } = {}): Promise<{
    events: {
      type: string
      contactId: string
      email: string
      timestamp: string
      data?: Record<string, unknown>
    }[]
    total: number
  }> {
    return this.client.get(`/campaigns/${id}/activity`, {
      page: options.page,
      limit: options.limit,
      eventType: options.eventType
    })
  }
}

File: sdks/python/mailat/resources/campaigns.py
"""Campaigns resource for managing marketing campaigns"""

from typing import TYPE_CHECKING, Any, Optional
from datetime import datetime
from ..types import Campaign, CampaignStats

if TYPE_CHECKING:
    from ..client import Mailat


class CampaignsResource:
    """Resource for managing marketing campaigns."""

    def __init__(self, client: "Mailat"):
        self._client = client

    def create(
        self,
        name: str,
        subject: str,
        list_ids: list[str],
        from_name: str,
        from_email: str,
        html_content: Optional[str] = None,
        text_content: Optional[str] = None,
        template_id: Optional[str] = None,
        reply_to: Optional[str] = None,
    ) -> Campaign:
        """Create a new campaign."""
        payload = {
            "name": name,
            "subject": subject,
            "listIds": list_ids,
            "fromName": from_name,
            "fromEmail": from_email,
        }
        if html_content:
            payload["htmlContent"] = html_content
        if text_content:
            payload["textContent"] = text_content
        if template_id:
            payload["templateId"] = template_id
        if reply_to:
            payload["replyTo"] = reply_to

        data = self._client.post("/campaigns", json=payload)
        return Campaign(**data)

    def get(self, campaign_id: str) -> Campaign:
        """Get campaign by ID."""
        data = self._client.get(f"/campaigns/{campaign_id}")
        return Campaign(**data)

    def update(
        self,
        campaign_id: str,
        name: Optional[str] = None,
        subject: Optional[str] = None,
        html_content: Optional[str] = None,
        text_content: Optional[str] = None,
    ) -> Campaign:
        """Update a campaign."""
        payload = {}
        if name:
            payload["name"] = name
        if subject:
            payload["subject"] = subject
        if html_content:
            payload["htmlContent"] = html_content
        if text_content:
            payload["textContent"] = text_content

        data = self._client.put(f"/campaigns/{campaign_id}", json=payload)
        return Campaign(**data)

    def delete(self, campaign_id: str) -> None:
        """Delete a campaign."""
        self._client.delete(f"/campaigns/{campaign_id}")

    def list(
        self,
        page: int = 1,
        limit: int = 50,
        status: Optional[str] = None,
    ) -> dict[str, Any]:
        """List campaigns."""
        params = {"page": page, "limit": limit}
        if status:
            params["status"] = status
        return self._client.get("/campaigns", params=params)

    def send(self, campaign_id: str) -> Campaign:
        """Send campaign immediately."""
        data = self._client.post(f"/campaigns/{campaign_id}/send")
        return Campaign(**data)

    def schedule(self, campaign_id: str, scheduled_at: datetime) -> Campaign:
        """Schedule campaign for future sending."""
        data = self._client.post(
            f"/campaigns/{campaign_id}/schedule",
            json={"scheduledAt": scheduled_at.isoformat()},
        )
        return Campaign(**data)

    def pause(self, campaign_id: str) -> Campaign:
        """Pause a sending campaign."""
        data = self._client.post(f"/campaigns/{campaign_id}/pause")
        return Campaign(**data)

    def resume(self, campaign_id: str) -> Campaign:
        """Resume a paused campaign."""
        data = self._client.post(f"/campaigns/{campaign_id}/resume")
        return Campaign(**data)

    def cancel(self, campaign_id: str) -> Campaign:
        """Cancel a scheduled or sending campaign."""
        data = self._client.post(f"/campaigns/{campaign_id}/cancel")
        return Campaign(**data)

    def get_stats(self, campaign_id: str) -> CampaignStats:
        """Get campaign statistics."""
        data = self._client.get(f"/campaigns/{campaign_id}/stats")
        return CampaignStats(**data)

    def preview(self, campaign_id: str, contact_id: Optional[str] = None) -> dict[str, str]:
        """Preview campaign HTML."""
        payload = {"contactId": contact_id} if contact_id else {}
        return self._client.post(f"/campaigns/{campaign_id}/preview", json=payload)

    def send_test(self, campaign_id: str, emails: list[str]) -> dict[str, int]:
        """Send test email for campaign."""
        return self._client.post(f"/campaigns/{campaign_id}/test", json={"emails": emails})
